### PR TITLE
For #1095. Create subtypes for JEXL AstNode.

### PR DESCRIPTION
**Pull request by dector**
Sealed classes seems like perfect solution for this issue.

### Pull Request checklist
- [x] **Quality**: This PR builds and passes detekt/ktlint checks (A pre-push hook is recommended)
- [x] **Tests**: This PR includes thorough tests or an explanation of why it does not
- [x] ~~**Changelog**: This PR includes [a changelog entry](https://github.com/mozilla-mobile/android-components/blob/master/docs/changelog.md) or does not need one~~
- [x] ~~**Accessibility**: The code in this PR follows [accessibility best practices](https://github.com/mozilla-mobile/shared-docs/blob/master/android/accessibility_guide.md) or does not include any user facing features~~

### CODE REVIEW

**Comment by pocmo** [comment]
Did a quick drive-by. Looks great so far! Looking forward to removing the properties from the base class. :)

**Comment by dector** [comment]
@pocmo Good. I'll continue refactoring then.

Unfortunately, simplifying `AstNode` will lead to many changes in parser as it's used heavily (e.g. as a partially-constructed object).

**Comment by dector** [comment]
Closing for now to not to trigger CI on each push.

Will squash/reorganize and reopen when there will be something more to discuss.

**Comment by dector** [comment]
Finished splitting to classes. Changed `toString()` implementation (I hope it's easier to follow now).

@pocmo it would be nice to hear you thoughts on it. The proposal is kinda-ready. I'll squash commits later (it's easier to follow/revert some changes with unsquashed ones).

**Comment by pocmo** [comment]
@dector We are currently cleaning up the list of PRs and I wonder what the state of this one here is? :)

**Comment by dector** [comment]
Hey. :)
I'm on vacation right now and will be back in few days. I'll continue
working on this soon.

**Comment by pocmo** [comment]
Enjoy your vacation! :)

**Comment by dector** [comment]
Thanks! I'm back now. Will rebase PR on master and prepare it for review till Tuesday I suppose.

**Comment by dector** [comment]
Squashed all commits. But I really want to add more real-life tests for parser. Is it possible to find them somewhere? :smiley_cat:

**Comment by pocmo** [comment]
Most tests and examples I took from the JavaScript and Python implementations/tests:
* https://github.com/mozilla/mozjexl
* https://github.com/mozilla/pyjexl

I guess you could also look at the coverage report and if there are uncovered sections write a test that uses exactly this part of the language. :)

**Comment by dector** [comment]
I was wondering why expressions like `!false` can't be evaluated. But checked on master branch - seems it works "as designed". :smile: 

I think we are ready to go with this.

File: components/lib/jexl/src/main/java/mozilla/components/lib/jexl/parser/StateMachine.kt [hunk 1]
```diff
@@ -141,7 +144,10 @@ internal val stateMachine: Map<State, StateMap> = mapOf(
     ),
     State.ARGUMENT_VALUE to StateMap(
         subHandler = { parser, node ->
-            parser.cursor!!.arguments.add(node!!)
+            val cursor = parser.cursor!!
+            if (cursor is Transformation) {
+                cursor.arguments.add(node!!)
+            }
```
Comment (dector) [inline components/lib/jexl/src/main/java/mozilla/components/lib/jexl/parser/StateMachine.kt]: @pocmo Maybe I should throw exception here if node is incorrect? What do you think?

Comment (dector) [reply components/lib/jexl/src/main/java/mozilla/components/lib/jexl/parser/StateMachine.kt]: Or even make cast that will throw exception.

File: components/lib/jexl/src/main/java/mozilla/components/lib/jexl/parser/Parser.kt [hunk 2]
```diff
@@ -108,7 +113,11 @@ internal class Parser(
         if (cursor == null) {
             tree = node
         } else {
-            cursor!!.right = node
+            cursor?.let { cursor ->
+                if (cursor is BranchNode) {
```
Comment (dector) [inline components/lib/jexl/src/main/java/mozilla/components/lib/jexl/parser/Parser.kt]: And cast here as well.

**Comment by pocmo** [comment]
> I was wondering why expressions like !false can't be evaluated. But checked on master branch - seems it works "as designed". 😄

Can you file this? Sounds like this *should* work. May be worth checking against mozjexl.

File: components/lib/jexl/src/main/java/mozilla/components/lib/jexl/parser/StateMachine.kt (hunk 1 above)
Comment (pocmo) [reply components/lib/jexl/src/main/java/mozilla/components/lib/jexl/parser/StateMachine.kt]: Yes, that sounds good - especially if this keeps the tests green. :)

File: components/lib/jexl/src/main/java/mozilla/components/lib/jexl/parser/Parser.kt (hunk 2 above)
Comment (pocmo) [reply components/lib/jexl/src/main/java/mozilla/components/lib/jexl/parser/Parser.kt]: 👍

File: components/lib/jexl/src/main/java/mozilla/components/lib/jexl/parser/StateMachine.kt (hunk 1 above)
Comment (dector) [reply components/lib/jexl/src/main/java/mozilla/components/lib/jexl/parser/StateMachine.kt]: :heavy_check_mark:

File: components/lib/jexl/src/main/java/mozilla/components/lib/jexl/parser/Parser.kt (hunk 2 above)
Comment (dector) [reply components/lib/jexl/src/main/java/mozilla/components/lib/jexl/parser/Parser.kt]: :heavy_check_mark:

**Comment by dector** [comment]
@pocmo.

There is no evaluator handler for `AstType.UNARY_EXPRESSION` now.
Will check with `mozjexl` and create follow up issue if result will be different.

**Comment by dector** [comment]
Follow-up issue: #3996

**Comment by dector** [comment]
@pocmo any updates on this?
Maybe I've got something wrong and something is expected from me? :)

**Pull request merged**